### PR TITLE
feature tiered subscription support: patch subscription endpoint

### PR DESCRIPTION
**Pull request by jackiemunroe**
Implement endpoint to allow for changing the Plan associated with a given Customer's Subscription

https://jira.mozilla.com/browse/COPS-861

### CODE REVIEW

File: src/sub/tests/unit/test_customer.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 )
 from sub.shared.db import SubHubAccountModel
 from sub.shared.cfg import CFG
-from sub.shared.exceptions import IntermittentError, ServerError
+from shared.exceptions import IntermittentError, ServerError
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_customer.py]: dropping `sub.shared` for `shared`

File: src/sub/tests/unit/test_exceptions.py [hunk 2]
```diff
@@ -73,7 +95,7 @@ def test_server_error():
     assert ex2.to_dict() == dict(message=message, some="payload")
 
 
-def test_SecretStringMissingError():
+def test_secret_string_missing_error():
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_exceptions.py]: this is unrelated to this change; also `SecretStringMissingError` is a specific name for the error class to be tested, therefore it is inappropriate to lowercase and separate by `_`.

File: src/sub/tests/unit/test_subhub.py [hunk 3]
```diff
@@ -51,15 +54,21 @@ def test_list_plans(mock_plans, mock_product, app):
     WHEN a request for plans is made
     THEN a success status of 200 is returned
     """
-    fh = open("src/sub/tests/unit/fixtures/stripe_plan_test1.json")
+    fh = open(
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]:
```
with open('filename') as f:
    f.read()
```

File: src/sub/tests/unit/test_subhub.py [hunk 4]
```diff
@@ -51,15 +54,21 @@ def test_list_plans(mock_plans, mock_product, app):
...
+        os.path.join(os.path.dirname(__file__), "fixtures/stripe_plan_test1.json")
+    )
     plan_test1 = json.loads(fh.read())
     fh.close()
 
-    fh = open("src/sub/tests/unit/fixtures/stripe_plan_test2.json")
+    fh = open(
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]: over

File: src/sub/tests/unit/test_subhub.py [hunk 5]
```diff
@@ -51,15 +54,21 @@ def test_list_plans(mock_plans, mock_product, app):
...
+        os.path.join(os.path.dirname(__file__), "fixtures/stripe_plan_test2.json")
+    )
     plan_test2 = json.loads(fh.read())
     fh.close()
 
-    fh = open("src/sub/tests/unit/fixtures/stripe_prod_test1.json")
+    fh = open(
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]: and over

File: src/sub/tests/unit/test_subhub.py [hunk 6]
```diff
@@ -178,25 +187,31 @@ def test_subscribe_success(
...
     mock_has_plan.return_value = False
     mock_fetch_customer.return_value = convert_to_stripe_object(cust_test1)
 
-    fh = open("src/sub/tests/unit/fixtures/stripe_sub_test1.json")
+    fh = open(os.path.join(os.path.dirname(__file__), "fixtures/stripe_sub_test1.json"))
     sub_test1 = json.loads(fh.read())
     fh.close()
-    fh = open("src/sub/tests/unit/fixtures/stripe_plan_test1.json")
+    fh = open(
+        os.path.join(os.path.dirname(__file__), "fixtures/stripe_plan_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]: also consider calculating the dirname of __file__ once and then using it; such that you lines aren't so long that black does this ugly wrapping

File: src/sub/tests/unit/test_subhub.py [hunk 7]
```diff
@@ -366,3 +387,420 @@ def test_plan_response_invalid(mock_plans, app):
...
+    def setUp(self) -> None:
+        sub_app = create_app()
+        with sub_app.app.app_context():
+            g.subhub_account = sub_app.app.subhub_account
+            g.subhub_deleted_users = sub_app.app.subhub_deleted_users
+
+        self.client = sub_app.app.test_client()
+
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_cust_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]: these have the `with` expressions; strange why the others do not; consistency for the win;  as before consider calculating the dirname of __file__ once and then use it in each; should keep these ugly black wraps from happening; also note that with file handles like this, it is very common to see `f` used not `fh`

File: src/sub/tests/unit/test_subhub.py [hunk 8]
```diff
@@ -366,3 +387,420 @@ def test_plan_response_invalid(mock_plans, app):
...
+    def test_plan_fetch_error(self):
+        self.get_subhub_user_mock.return_value = self.subhub_user
+        self.retrieve_customer_mock.return_value = self.customer
+        self.retrieve_plan_mock.side_effect = stripe.error.InvalidRequestError(
+            "message", param="param"
+        )
+
+        data = {"plan_id": "plan_test2"}
+
+        response = self.client.patch(
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_subhub.py]: this block is repeated over and over, could be a good candidate for a function that you just pass the `data` to because it looks like everything else is the same every time

File: src/sub/tests/unit/test_vendor.py [hunk 9]
```diff
@@ -1,114 +1,441 @@
...
 from sub.shared import vendor, utils
+from shared.exceptions import EntityNotFoundError
 
 
-def disable_base():
-    stripe.api_base = "http://example.com"
+class TestStripeCustomerCalls(TestCase):
+    def setUp(self):
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_cust_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_vendor.py]: same thing, calculcate dirname and avoid ugly black wrapping

File: src/sub/tests/unit/test_vendor.py [hunk 10]
```diff
@@ -1,114 +1,441 @@
...
+        self.mock_list_subscription.side_effect = APIError("message")
+
+        with self.assertRaises(APIError):
+            vendor.list_customer_subscriptions(cust_id="cust_123")
+
+
+class TestStripeChargeCalls(TestCase):
+    def setUp(self) -> None:
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_ch_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_vendor.py]: the thing

File: src/sub/tests/unit/test_vendor.py [hunk 11]
```diff
@@ -1,114 +1,441 @@
...
+        self.retrieve_charge_mock.side_effect = APIError("message")
+
+        with self.assertRaises(APIError):
+            vendor.retrieve_stripe_charge("in_test1")
+
+
+class TestStripeInvoiceCalls(TestCase):
+    def setUp(self) -> None:
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_in_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_vendor.py]: the thing

File: src/sub/tests/unit/test_vendor.py [hunk 12]
```diff
@@ -1,114 +1,441 @@
...
+        self.retrieve_invoice_mock.side_effect = APIError("message")
+
+        with self.assertRaises(APIError):
+            vendor.retrieve_stripe_invoice("in_test1")
+
+
+class TestStripePlanCalls(TestCase):
+    def setUp(self) -> None:
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_plan_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_vendor.py]: again

File: src/sub/tests/unit/test_vendor.py [hunk 13]
```diff
@@ -1,114 +1,441 @@
...
+        )
+
+        with self.assertRaises(InvalidRequestError) as e:
+            vendor.find_stripe_plan("plan_test1")
+
+
+class TestStripeProductCalls(TestCase):
+    def setUp(self) -> None:
+        with open(
+            os.path.join(os.path.dirname(__file__), "fixtures/stripe_prod_test1.json")
```
Comment (sidler-mozilla) [inline src/sub/tests/unit/test_vendor.py]: !

File: src/sub/tests/unit/test_customer.py (hunk 1 above)
Comment (marty331) [reply src/sub/tests/unit/test_customer.py]: resolved.

File: src/sub/tests/unit/test_subhub.py (hunk 3 above)
Comment (marty331) [reply src/sub/tests/unit/test_subhub.py]: reworked and resolved.

File: src/sub/tests/unit/test_subhub.py (hunk 6 above)
Comment (marty331) [reply src/sub/tests/unit/test_subhub.py]: Done (good suggestion)

File: src/sub/tests/unit/test_vendor.py (hunk 9 above)
Comment (marty331) [reply src/sub/tests/unit/test_vendor.py]: done

File: src/sub/tests/unit/test_vendor.py (hunk 10 above)
Comment (marty331) [reply src/sub/tests/unit/test_vendor.py]: is done

File: src/sub/tests/unit/test_vendor.py (hunk 11 above)
Comment (marty331) [reply src/sub/tests/unit/test_vendor.py]: is done

File: src/sub/tests/unit/test_vendor.py (hunk 12 above)
Comment (marty331) [reply src/sub/tests/unit/test_vendor.py]: done

File: src/sub/tests/unit/test_vendor.py (hunk 13 above)
Comment (marty331) [reply src/sub/tests/unit/test_vendor.py]: done

File: src/sub/tests/unit/test_subhub.py (hunk 8 above)
Comment (marty331) [reply src/sub/tests/unit/test_subhub.py]: Unique tests.

File: src/sub/tests/unit/test_subhub.py (hunk 4 above)
Comment (marty331) [reply src/sub/tests/unit/test_subhub.py]: done

File: src/sub/tests/unit/test_exceptions.py (hunk 2 above)
Comment (marty331) [reply src/sub/tests/unit/test_exceptions.py]: done

File: src/sub/swagger.yaml [hunk 14]
```diff
@@ -301,6 +301,51 @@ paths:
       parameters:
         - $ref: '#/parameters/uidParam'
         - $ref: '#/parameters/subIdParam'
+    patch:
+      operationId: sub.payments.update_subscription
+      tags:
+        - Subscription
```
Comment (marty331) [inline src/sub/swagger.yaml]: Should be Subscriptions otherwise it is on a new node.

Comment (jackiemunroe) [reply src/sub/swagger.yaml]: fixed :)
